### DISCUSSION TITLE
How to override an environment variable via the CLI?

### DISCUSSION THREAD

**Discussion by daniel-woodhead** [AnWeber/httpyac#773] [General]
Given a `.httpyac.js` file:
```
module.exports = {
  environments: {
    $default: {},
    $shared: {},
    dev: {
      baseUrl: "https://httpbin.org",
      param: "foo",
    },
  },
};
```

And a `test.http` file:
```
GET {{baseUrl}}/get?param={{param}} HTTP/1.1
```

How can I use the `dev` environment but override one or more variables?

This does not seem to work:
```
httpyac send test.http --all --env dev --var param=bar
```

**Comment by alekdavis** [comment]
@daniel-woodhead Can't you use environment files? Not really sure what your use case is about. Which variable do you want to override?

**Reply by daniel-woodhead** [reply]
My use case was with a secret variable. I wanted to define the variable in `.httpyac.js` as an empty string so that A) the value is not committed but B) it's clear that the variable exists and that a value needs to be provided:
```
module.exports = {
  environments: {
    dev: {
      secret: "",
    },
  },
};
```

Then I wanted to be able to provide the value in CI like so:
```
httpyac send test.http --all --env dev --var secret=...
```

**Reply by alekdavis** [reply]
You did not answer my question. Why don't you use environment file(s)? The httpYac. js is the project settings file. It is not intended for environment variables, so I am a bit confused why you want to use it when the capability exists using the environment files.

**Reply by AnWeber** [reply]
@daniel-woodhead Whether you use .httpyac.js, .env or Intellij Config Format should not really matter. These are actually just different types of configuration and as long as you don't mix them, it shouldn't matter. I would actually have defined my own configuration sequence, but I have now violated it (what you don't test will break at some point). In other words, I think the following sequence makes sense. And I will fix this. Currently Variables defined per CLI Args is overriden by environments, but this is not intentional)

```
// stick with one file format between 1-3.
1. Environment Variables defined in .httpyac.js, jttpyac.config.js, ...
2. .env Variables
3. Intellij Environment Variables

4. Variables defined per CLI Args (allow simple override of environment settings)
5. Variables defined in http Files
```
https://github.com/AnWeber/httpyac/commit/f7c692bf5eca80d3cc2df392099f6fb703d707a9

**Reply by alekdavis** [reply]
@daniel-woodhead Basically, if you are trying to prevent sensitive values from being checked into source control, but want to keep some kind of structure in the environment file to keep a placeholder for the sensitive value, it is supported by both dotenv and IntelliJ environments. I use IntelliJ environments, and this is my sample: https://github.com/alekdavis/httpyac-extras/tree/main/env. What you need is just follow the instructions at https://github.com/alekdavis/httpyac-extras/tree/main/env#http-clientprivatesampleenvjson-file. And keep the `.gitignore` file settings from that repo to prevent the private file from being checked in. When using CLI, you just specify the value using the `--var` switch.

**Reply by daniel-woodhead** [reply]
Thank you both. @alekdavis not using environment files was just a personal choice. The docs say using .httpyac.js is supported for this (https://httpyac.github.io/guide/environments.html#json) and I liked the idea of having all settings defined in a single file. IntelliJ environments look like a good alternative though so thanks for the suggestion.
